route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,190 +41,210 @@ namespace route_following_plugin
+        }
 
-            ++last_lanelet_index;
+        lanelet::BasicPoint2d current_loc(pose_msg_.pose.position.x, pose_msg_.pose.position.y);
+        
+        double current_downtrack = wm_->routeTrackPos(current_loc).downtrack;
+        double speed_progress = current_speed_;
+        
+        //update start distance of first maneuver and time for all maneuvers
+        setManeuverStartDist(latest_maneuver_plan_.front(), current_downtrack);
+        //Update time progress for maneuvers
+        updateTimeProgress(latest_maneuver_plan_, ros::Time::now());
+        //update starting speed
+        
+        //Return the set of maneuvers which intersect those 15s
+        int i=0;
+        double planned_time = 0.0;
+        while(planned_time < min_mvr_duration_ && i < latest_maneuver_plan_.size()){
+            double maneuver_start_time = GET_MANEUVER_PROPERTY(latest_maneuver_plan_[i], start_time).toSec();
+            double maneuver_end_time = GET_MANEUVER_PROPERTY(latest_maneuver_plan_[i],end_time).toSec();
+            planned_time += maneuver_end_time - maneuver_start_time;
 
+            resp.new_plan.maneuvers.push_back(latest_maneuver_plan_[i]);
+            i++;
         }
-        resp.new_plan.maneuvers.push_back(
-            composeStopandWaitManeuverMessage(current_progress,total_maneuver_length,
-            speed_progress,shortest_path[last_lanelet_index].id(),
-            shortest_path[last_lanelet_index].id(),time_progress,time_req_to_stop));
         
-        ROS_DEBUG_STREAM("Done Loop: approaching_route_end: " << approaching_route_end);
-        if(approaching_route_end){
-            ros::Time end_time = ros::Time::now();
-            if (resp.new_plan.maneuvers.size() > 0) {
-                end_time = resp.new_plan.maneuvers.back().lane_following_maneuver.end_time;
-            }
-            
-            resp.new_plan.maneuvers.push_back(
-                composeStopandWaitManeuverMessage(current_progress,total_maneuver_length,
-                speed_progress,shortest_path[last_lanelet_index].id(),
-                shortest_path[last_lanelet_index].id(),end_time,time_req_to_stop)); 
-        }
         if(resp.new_plan.maneuvers.size() == 0)
         {
             ROS_WARN_STREAM("Cannot plan maneuver because no route is found");
         }
         return true;
     }
-    void RouteFollowingPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
-    {
-        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+
+    void RouteFollowingPlugin::updateTimeProgress(std::vector<cav_msgs::Maneuver>& maneuvers, ros::Time start_time){
+        ros::Time time_progress = start_time;
+        ros::Time prev_time = time_progress;
+
+        for(auto maneuver : maneuvers){
+            ros::Duration maneuver_duration =GET_MANEUVER_PROPERTY(maneuver,end_time) - GET_MANEUVER_PROPERTY(maneuver, start_time);
+            time_progress += maneuver_duration;
+            switch(maneuver.type){
```

Comment:
If the class structure was refactored to put these time params in a base class, then this whole switch statement could go away and you could just say maneuver.start_time = ..., for example.  Would simplify other areas of the code as well.  Something for a future enhancement.